modules/streams/src/main/scala/dev/profunktor/redis4cats/pubsub/internals/Subscriber.scala

```diff
@@ -18,151 +18,435 @@ package dev.profunktor.redis4cats
-      key: K
-  ): F[Unit] = subs.evalUpdate { subscribers =>
-    subscribers.get(key) match {
-      case None =>
-        Log[F]
-          .error(
-            s"We were notified about stream termination for $key but we don't have a subscription, " +
-              s"this is a bug in redis4cats!"
-          )
-          .as(subscribers)
-      case Some(sub) =>
-        if (!sub.isLastSubscriber) subscribers.updated(key, sub.removeSubscriber).pure
-        else sub.cleanup.as(subscribers - key)
-    }
+
+private[pubsub] object Subscriber {
+
+  def make[F[_]: Async: FutureLift: Log, K, V](
+      subConnection: StatefulRedisPubSubConnection[K, V]
+  ): Resource[F, SubscribeCommands[F, Stream[F, *], K, V]] =
+    for {
+      state <- Resource.eval(
+                 State.fromMapRefs[F, K, V](
+                   channelCommands = SubscriptionCommands.channel(subConnection),
+                   patternCommands = SubscriptionCommands.pattern(subConnection)
+                 )
+               )
+      // We only use a single listener for all channels and patterns.
+      // Since we have a map of all subscriptions, we can dispatch messages to
+      // the right topic directly.
+      // Lettuce calls the listeners one by one (for every subscribe,
+      // unsubscribe, message, ...), so using multiple listeners when we can
+      // find the right subscription easily, is inefficient.
+      dispatcher <- Dispatcher.sequential[F] // we have no parallelism in the listener
+      _ <- Resource.make {
+             val listener = State.listener(state, dispatcher)
+             Sync[F].delay(subConnection.addListener(listener)).as(listener)
+           }(listener => Sync[F].delay(subConnection.removeListener(listener)))
+    } yield new Subscriber(state)
+
+  private final case class State[F[_], K, V](
+      channelSubs: SubscriptionMap[F, RedisChannel[K], V],
+      patternSubs: SubscriptionMap[F, RedisPattern[K], RedisPatternEvent[K, V]]
+  )
+
+  private object State {
+    def fromMapRefs[F[_]: Async: Log, K, V](
+        channelCommands: SubscriptionCommands[F, RedisChannel[K]],
+        patternCommands: SubscriptionCommands[F, RedisPattern[K]]
+    ): F[State[F, K, V]] =
+      (
+        SubscriptionMap.makeMapRef[F, RedisChannel[K], V](channelCommands),
+        SubscriptionMap.makeMapRef[F, RedisPattern[K], RedisPatternEvent[K, V]](patternCommands)
+      ).mapN(apply)
+
+    // Note that this blocks when `onMessage` semantically blocks.
+    // The current `State` implementation using `Topic` semantically blocks when
+    // one of the subscribers of the `Topic` is behind.
+    def listener[F[_], K, V](
+        state: State[F, K, V],
+        dispatcher: Dispatcher[F]
+    ): RedisPubSubListener[K, V] =
+      new RedisPubSubAdapter[K, V] {
+        override def message(ch: K, msg: V): Unit =
+          try
+            dispatcher.unsafeRunSync(state.channelSubs.onMessage(RedisChannel(ch), msg))
```

Comment:
You mention this above, but since we funnel all messages through a single sequential dispatcher, which is backed by a single queue for the IO effects, this could create "head-of-line" blocking. There could exist a topic with a single subscriber blocking, and that would lead to block subscribers on all other topics. And this is b/c `onMessage` just routes to a topic, then publishes to it.

I think solutions would be to add queues or do routing at the Java layer.

Reply:
> this could create "head-of-line" blocking

We already have head of line blocking, lettuce using a single connection to redis and calls listeners one by one. A topic with a single blocking subscriber currently also already blocks all subscriptions. This PR doesn't solve that issue no. The main change in this PR is to avoid blocking while subscribing and unsubscribing, the rest of the behaviour should have stayed the same.

I don't think it is an issue that we can solve, I think the best we can do is to document it clearly that blocked subscribers can block all redis subscriptions and that clients should handle this on their end.